articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,234 @@
+
+
+## Runtime
+
+TODO: Some transition paragraph here about ros launch
+
+### Unqualified context path
+
+For nodes with unqualified context paths, the context directory will subsequently default to the root level context.
+
+``` xml
+<launch>
+  <node pkg="demo_nodes_cpp" exec="talker"/>
+  <node pkg="demo_nodes_cpp" exec="listener"/>
+</launch>
+```
+
+```
+$ tree contexts/
+contexts/
+├── cert.pem
+├── governance.p7s
+├── identity_ca.cert.pem -> ../public/identity_ca.cert.pem
+├── key.pem
+├── permissions_ca.cert.pem -> ../public/permissions_ca.cert.pem
+└── permissions.p7s
+```
+
+### Pushed unqualified context path
+
+For nodes with unqualified context paths pushed by a namespace, the context directory will subsequently be pushed to the relative sub-folder.
+
+``` xml
+<launch>
+  <node pkg="demo_nodes_cpp" exec="talker"/>
+  <group>
+    <push_ros_namespace namespace="foo"/>
+    <node pkg="demo_nodes_cpp" exec="listener"/>
+  </group>
+</launch>
+```
+
+```
+$ tree --dirsfirst contexts/
+contexts/
+├── foo
+│   ├── cert.pem
+│   ├── governance.p7s
+│   ├── identity_ca.cert.pem
+│   ├── key.pem
+│   ├── permissions_ca.cert.pem
+│   └── permissions.p7s
+├── cert.pem
+├── governance.p7s
+├── identity_ca.cert.pem
+├── key.pem
+├── permissions_ca.cert.pem
+└── permissions.p7s
+```
+> Symbolic links suppressed for readability
+
+### Relatively pushed qualified context path
+
+For nodes with qualified context paths pushed by a namespace, the qualified context directory will subsequently be pushed to the relative sub-folder.
+
+``` xml
+<launch>
+  <group>
+    <push_ros_namespace namespace="foo"/>
+    <node pkg="demo_nodes_cpp" exec="listener" context="bar"/>
+  </group>
+</launch>
+```
+
+```
+$ tree --dirsfirst contexts/
+contexts/
+└── foo
+    └── bar
+        ├── cert.pem
+        ├── governance.p7s
+        ├── identity_ca.cert.pem
+        ├── key.pem
+        ├── permissions_ca.cert.pem
+        └── permissions.p7s
+```
+
+### Fully qualified context path
+
+For nodes with fully qualified context paths, namespacs do not subsequently push the relative sub-folder.
+
+``` xml
+<launch>
+  <group>
+    <push_ros_namespace namespace="foo"/>
+    <node pkg="demo_nodes_cpp" exec="listener" context="/bar"/>
+  </group>
+</launch>
+```
+
+```
+$ tree --dirsfirst contexts/
+contexts/
+└── bar
+    ├── cert.pem
+    ├── governance.p7s
+    ├── identity_ca.cert.pem
+    ├── key.pem
+    ├── permissions_ca.cert.pem
+    └── permissions.p7s
+```
+
+
+## Alternatives
+
+### Context path orthogonal to namespace
+
+An alternative to reusing namespaces to hint the context path could be to completely disassociate the two entirely, treating the context path as it's own unique identifier. However, having to book keep both identifier spaces simulations may introduce to many degrees of freedom that a human could groc or easily introspect via tooling.
```

Comment:
From the offline notes, we visited the idea of forcing users to manually specify context namespaces, as to be comely orthogonal to node namespaces, but concluded that tracking two orthogonal namespaces would be a large mental overhead for users, as well as require much more boilerplate.

> In the case of launchfiles: how to specify what gets in what context?
Could force people to provide context attributes everywhere ?
Inconvenient and can duplicate info. Would need "namespace context" but it gets hairy and duality with nodes name and namespaces

Thus we iterated further on the idea of allowing the existing namespace elements in lauchfiles to hint/push the relative context path. However, the user could still override the relative context path using a absolute context path for the node element's context attribute, just like with topics/services/actions:

> If context specified use context-prefix/context otherwise can be overridden using absolute context

Perhaps we could also add these additional observations to the alternative section.